packages/wdio-browserstack-service/src/instrumentation/funnelInstrumentation.ts

```diff
@@ -151,3 +154,57 @@ function getReferrer(framework?: string) {
     const fullName = framework ? 'WebdriverIO-' + framework : 'WebdriverIO'
     return `${fullName}/${BSTACK_SERVICE_VERSION}`
 }
+
+async function sendTcgDownError(config: BrowserStackConfig) {
+    await fireFunnelTestEvent('SDKTestTcgDownResponse', config)
+}
+
+async function sendInvalidTcgAuthResponse(config: BrowserStackConfig) {
+    await fireFunnelTestEvent('SDKTestInvalidTcgAuthResponseWithUserImpact', config)
+}
+
+async function sendTcgAuthFailure(config: BrowserStackConfig) {
+    await fireFunnelTestEvent('SDKTestTcgAuthFailure', config)
+}
+
+async function sendTcgtInitSuccessful(config: BrowserStackConfig) {
+    await fireFunnelTestEvent('SDKTestTcgtInitSuccessful', config)
+}
+
+async function sendInitFailedResponse(config: BrowserStackConfig) {
+    await fireFunnelTestEvent('SDKTestInitFailedResponse', config)
+}
+
+async function sendTcgProxyFailure(config: BrowserStackConfig) {
+    await fireFunnelTestEvent('SDKTestTcgProxyFailure', config)
+}
+
+export function handleHealingInstrumentation (
```

Comment:
Please consider simplifying the method. Example below.

```
export function handleHealingInstrumentation (
    authResult: BrowserstackHealing.InitErrorResponse | BrowserstackHealing.InitSuccessResponse,
    config: BrowserStackConfig,
    isSelfHealEnabled: boolean | undefined,
    options: BrowserstackConfig & Options.Testrunner
) {
    const { message, isAuthenticated, status, userId, isHealingEnabled } = authResult as any;
    const proxyHost = config.proxyHost || config.localProxyHost || options.proxyHost || options.localProxyHost;

    if (isSelfHealEnabled) {
        if (message === 'Upgrade required') {
            BStackLogger.warn('Please upgrade Browserstack Service to the latest version to use the self-healing feature.');
            return;
        }

        if (!isAuthenticated) {
            if (status === 503) {
                BStackLogger.warn('Something went wrong. Disabling healing for this session. Please try again later.');
                sendTcgDownError(config);
            } else {
                BStackLogger.warn('Authentication Failed. Disabling Healing for this session.');
                sendTcgAuthFailure(config);
            }
            return;
        }

        if (isAuthenticated) {
            if (!isHealingEnabled) {
                BStackLogger.warn('Healing is not enabled for your group, please contact the admin');
            } else if (userId) {
                sendTcgtInitSuccessful(config);
            }
            return;
        }
    }

    if (status >= 400) {
        sendInitFailedResponse(config);
    } else if (!status) {
        sendInvalidTcgAuthResponse(config);
    } else if (proxyHost) {
        // TODO: Might need to explore more on proxy handling and modify this condition accordingly
        sendTcgProxyFailure(config);
    }
}
```

Reply:
Updated